### PR TITLE
refactor(plugin/vite-resolve): use oxc-resolver to resolve optional peer dep value from package.json

### PR DESCRIPTION
**Pull request by sapphi-red**
### Description

Removed the custom package.json resolving logic and use oxc-resolver for that (adds a cache too).

### CODE REVIEW

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs [hunk 1]
```diff
@@ -1,24 +1,99 @@
...
+};
 
 use dashmap::DashMap;
 use rolldown_common::PackageJson;
+use serde::{de::IgnoredAny, Deserialize};
 
 #[derive(Debug, Default)]
 pub struct PackageJsonCache {
-  package_json_cache: DashMap<PathBuf, Arc<PackageJson>>,
+  side_effects_cache: DashMap<PathBuf, Arc<PackageJson>>,
```
Comment (Boshen) [inline crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Can you use fx hash for all these hashmaps? e.g. https://github.com/oxc-project/oxc-resolver/blob/fe70f41f0dd1a7d0cfb854839c0b1110cbad7a35/src/cache.rs#L30

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs [hunk 2]
```diff
@@ -1,24 +1,101 @@
...
+      self.optional_peer_dep_cache.insert(oxc_pkg_json.realpath.clone(), Arc::clone(&pkg_json));
+      pkg_json
+    }
+  }
+}
+
+#[derive(Debug, Default)]
+pub struct PackageJsonWithOptionalPeerDependencies {
+  pub name: String,
+  pub optional_peer_dependencies: BTreeSet<String>,
```
Comment (Boshen) [inline crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: One more here.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs [hunk 3]
```diff
@@ -1,24 +1,101 @@
...
+        .collect(),
+    })
+  }
+}
+
+#[derive(Deserialize)]
+struct PackageJsonWithPeerDependenciesRaw {
+  pub name: String,
+  #[serde(rename = "peerDependencies")]
+  pub peer_dependencies: Option<BTreeMap<String, IgnoredAny>>,
```
Comment (Boshen) [inline crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Another one.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs [hunk 4]
```diff
@@ -1,24 +1,101 @@
...
+  #[serde(rename = "peerDependenciesMeta")]
+  pub peer_dependencies_meta: Option<BTreeMap<String, PackageJsonPeerDependenciesMetaRaw>>,
```
Comment (Boshen) [inline crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Another one.

**Review by Boshen — COMMENTED** [review]
Does this plugin have access to an `OxcResolver` instance?

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 1 above)
Comment (sapphi-red) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Made #2971 for other places that wasn't using fx hash.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 4 above)
Comment (sapphi-red) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Ideally, I wanted these two to be a Map backed by a Vec as I don't need the key access. I don't know if `HashMap` has smaller cost than `BTreeMap` or not. If that's the case, I'll replace these with a HashMap.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 2 above)
Comment (sapphi-red) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Do you think it's better to use a HashSet here? The size of this set should be small and will only looked up in minor cases.
I'm not sure whether which of `BTreeSet` or `HashSet` or a sorted `Vec`+`binary_search` is better.

**Comment by sapphi-red** [comment]
> Does this plugin have access to an `OxcResolver` instance?

I think no, if you mean the `OxcResolver` instance in rolldown's internal resolver.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 4 above)
Comment (sapphi-red) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Ah, actually I need the key access for `peer_dependencies_meta`.

Comment (Boshen) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: `HashMap` should perf better in this case.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 2 above)
Comment (Boshen) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: `HashMap` is always cheaper.

File: crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs (hunk 4 above)
Comment (IWANABETHATGUY) [reply crates/rolldown_plugin_vite_resolve/src/package_json_cache.rs]: Unless you needs visit the key in order, hashmap should outperform in most scenarios.

**Review by hyf0 — APPROVED** [review]
No blocking issues. Let's merge it first.

**Pull request merged**